docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,698 @@
+// included in a data square. It also has methods to help progressively count
+// how many shares the messages written take up.
+type MessageShareSplitter struct {
+  shares [][]NamespacedShare
+  count  int
+}
+...
+// WriteNamespacedPaddedShares adds empty shares using the namespace of the last written share.
+// This is useful to follow the message layout rules. It assumes that at least
+// one share has already been written, if not it panics.
+func (mss *MessageShareSplitter) WriteNamespacedPaddedShares(paddedShares int) {
+  if len(mss.shares) == 0 {
+      panic("Cannot write empty namespaced shares on an empty MessageShareSplitter")
+  }
+  if paddedShares == 0 {
+      return
+  }
+  lastMessage := mss.shares[len(mss.shares)-1]
+  mss.shares = append(mss.shares, namespacedPaddedShares(lastMessage[0].ID, paddedShares))
+  mss.count += paddedShares
+}
+```
+
+Now we simply combine this new functionality with the `share_index`s described above, and we can properly split and pad messages when needed. Note, the below implementation allows indexes to not be used. This is important, as it allows the same implementation to be used in the cases where we don't want to split messages using wrapped transactions, such as supporting older networks or when users create commitments to sign over for `MsgWirePayForData`
+
+```go
+func SplitMessages(cursor int, indexes []uint32, msgs []coretypes.Message, useShareIndexes bool) ([][]byte, error) {
+  if useShareIndexes && len(indexes) != len(msgs) {
+      return nil, ErrIncorrectNumberOfIndexes
+  }
+  writer := NewMessageShareSplitter()
+  for i, msg := range msgs {
+      writer.Write(msg)
+      if useShareIndexes && len(indexes) > i+1 {
+          paddedShareCount := int(indexes[i+1]) - (writer.Count() + cursor)
+          writer.WriteNamespacedPaddedShares(paddedShareCount)
+      }
+  }
+  return writer.Export().RawShares(), nil
+}
+```
+
+When parsing the message shares, we can simply ignore these added namespaced padded shares.
+
+### Implement the non-interactive default logic
+
+recall our non-interactive default message layout rule:
+
+> **Messages must begin at a location aligned with the largest power of 2 that is not larger than the message length or k.**
+
+The key to arranging the square into non-interactive defaults is calculating then next "aligned power of 2". We do that here statelessly with two simple functions.
+
+```go
+// NextAlignedPowerOfTwo calculates the next index in a row that is an aligned
+// power of two and returns false if the entire the msg cannot fit on the given
+// row at the next aligned power of two. An aligned power of two means that the
+// largest power of two that fits entirely in the msg or the square size. pls
+// see specs for further details. Assumes that cursor < k, all args are non
+// negative, and that k is a power of two.
+// https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#non-interactive-default-rules
```

Comment:
Do we really want to link to spec here and in the code? If yes, then please use permalinks.

Reply:
changed here, but created an issue to change the rest of the code https://github.com/celestiaorg/celestia-app/issues/780